grails-app/migrations/views/cycle-count-session.sql

```diff
@@ -1,40 +1,74 @@
+
+    # Stuff we need to include now or wait for the materialized view
+    -- Negative item count
+    (SELECT SUM(CASE WHEN pai.quantity_on_hand < 0 THEN 1 ELSE 0 END)
+     FROM product_availability pai
+     WHERE pai.product_id = product_availability.product_id
+       AND pai.location_id = product_availability.location_id)                   as negative_item_count,
+
+    -- Date last counted
+    (select date_counted
+     from product_count_history
+     where product_count_history.product_id = product_availability.product_id
+       and product_count_history.inventory_id = location.inventory_id) as date_last_count,
+
+    NULL                                                                         as date_next_count,
+    NULL                                                                         as date_latest_inventory
+FROM product_availability
+         JOIN location on product_availability.location_id = location.id
+         JOIN product ON product_availability.product_id = product.id
+    # FIXME need to add a unique constraint to prevent duplicate inventory levels for a single facility / product
+    # Using min(abc_class) means we'll return the highest value if there are multiple inventory levels with different abc classes (A, B, C)
+        LEFT OUTER JOIN (select
+                              inventory_id,
+                              product_id,
+                              min(abc_class) as abc_class
+                          from inventory_level where inventory_level.internal_location_id IS NULL
+                          group by inventory_id, product_id) as inventory_level_summary
+                         ON product_availability.product_id = inventory_level_summary.product_id AND location.inventory_id = inventory_level_summary.inventory_id
+
+        LEFT OUTER JOIN (select cycle_count_request.facility_id                          as facility_id,
+                                cycle_count_request.product_id                           as product_id,
+                                cycle_count_request.id                                   as cycle_count_request_id,
+                                COALESCE(cycle_count.status, cycle_count_request.status) as status
+                         from cycle_count_request
+                                  left outer join cycle_count on cycle_count_request.cycle_count_id = cycle_count.id
+                         where cycle_count_request.id IS NULL
+                            OR (cycle_count_request.status <> 'COMPLETED'
+                                    AND cycle_count_request.status <> 'CANCELED')) as cycle_count_request_summary
+            on cycle_count_request_summary.product_id = product_availability.product_id
+                   and cycle_count_request_summary.facility_id = product_availability.location_id
+
+
+GROUP BY product_availability.location_id, product_availability.product_id
+HAVING sum(product_availability.quantity_on_hand) > 0
+ORDER BY NULL
```

Comment:
To be clear, there is a purpose to this line. Since I'm trying to tune performance as much as possible I decided to remove the default sorting on the view and leave it to the query to decide.  

I read about this in the following stackoverflow post (mysql reference included)
https://dba.stackexchange.com/questions/208166/group-by-needs-order-by-null-to-avoid-filesort

Reply:
Additional resources
* https://dba.stackexchange.com/questions/202386/mariadb-why-order-by-null-command-doesnt-remain-using-filesort-on-the-extr
* https://dev.mysql.com/doc/refman/5.7/en/order-by-optimization.html

Reply:
As I wrote in my review this `HAVING sum(product_availability.quantity_on_hand) > 0` seems risky, as looks like we can have negative quantities, so I think it might remove some records. Or if one row has a high negative value and the other one has a low positive value, then should it be removed here eitherway?

Reply:
I would be fine removing it but just note that there's a requirement to return products with quantity on hand > 0. My point was for someone to investigate that requirement to help us interpret its intent. 

To be honest, I actually think this clause needs to be in the query against cycle count session, not in the view itself. But if the requirement states that we should ignore inventory items (not products) with quantity on hand <= 0 then the original view definition is "correct". It would just produce inaccurate aggregate data so it didn't make sense to me.